#### Instruction: Part 1 - Data Preparation and Exploration - Total Points: 20
For this part, you will be using Capitaliq.com to download the dataset as an excel file and clean it on Jupyter notebooks.

1.1 Download the constituents of the S&P 1500 using Capital IQ. 

1.2 Clean and organize the Excel file that you acquire from Capital IQ. Create a pandas data frame and save the clean file as sp1500.csv

1.3 Determine how many constituents and industries are part of the S&P 1500. Before dropping NAs and after dropping NAs.

1.4 Determine the top 5 worst performing industries and top 5 best performing industries.

1.5 Create a new data frame for the worst industry and a new data frame for the best industry both based on % Price Change [YTD]. For each of these new data frames, use the % Price Change [30 Day] to plot the top five ranking for worst and best performance constituents.

Submit a clean and organized notebook and the corresponding dataset answering the questions for this part.

# Advanced Data Science &amp; Python for Finance  <br><br> Final Project <br><br> Part 1 - Data Preparation and Exploration
-----

FIN580-59305 Spring 2020

Name: Danni Chen <br>
NetID: dannic3

Master of Science in Finance, University of Illinois at Urbana-Champaign.

* https://www.linkedin.com/in/dannie-chen-dc/

## Packages and Settings

In [183]:
import pandas as pd

## 1.1 Download and Read the SP1500 Constituents Data
#### Capital IQ - S&P Composite 1500 (^SP1500) > Constituents

In [184]:
sp_df = pd.read_excel("../../data/SP1500.xls", skiprows = 14, skipfooter = 12)

In [185]:
sp_df.head()

Company Name Exchange:Ticker Currency  \
0   3D Systems Corporation        NYSE:DDD      USD   
1               3M Company        NYSE:MMM      USD   
2                8x8, Inc.       NYSE:EGHT      USD   
3  A. O. Smith Corporation        NYSE:AOS      USD   
4               AAON, Inc.   NasdaqGS:AAON      USD   

   Market Cap (mm) [Latest]*† Revenue (mm) % Price Change [Last Day]  \
0                      923.60       629.09                   (8.48%)   
1                    85474.18        32348                   (2.19%)   
2                     1628.06       418.53                   (6.31%)   
3                     6747.51       2992.7                   (1.72%)   
4                     2363.66       469.33                   (4.68%)   

  % Price Change [30 Day] % Price Change [YTD] % Price Change [12 Month]  \
0                  17.37%             (11.20%)                  (26.91%)   
1                  11.61%             (15.77%)                  (20.14%)   
2                  23.47%             (13.17%)                  (32.12%)   
3                  16.96%             (12.57%)                  (20.67%)   
4                 (3.14%)              (8.10%)                   (8.94%)   

   Price Close   P/E*† P/BV*†                              Primary Industry  
0         7.77      NM  1.75x  Technology Hardware, Storage and Peripherals  
1       148.60  17.47x  8.42x                      Industrial Conglomerates  
2        15.89      NM  7.30x                          Application Software  
3        41.65  18.75x  4.06x                             Building Products  
4        45.41  44.47x  8.15x                             Building Products

In [138]:
sp_df.tail()

Company Name Exchange:Ticker Currency  \
1501              Zebra Technologies Corporation   NasdaqGS:ZBRA      USD   
1502                Zimmer Biomet Holdings, Inc.        NYSE:ZBH      USD   
1503  Zions Bancorporation, National Association   NasdaqGS:ZION      USD   
1504                                 Zoetis Inc.        NYSE:ZTS      USD   
1505                                 Zumiez Inc.   NasdaqGS:ZUMZ      USD   

      Market Cap (mm) [Latest]*† Revenue (mm) % Price Change [Last Day]  \
1501                    11974.80         4471                   (1.79%)   
1502                    24026.74       7982.2                   (2.88%)   
1503                     5063.03         2533                   (2.25%)   
1504                    60565.33         6260                   (1.38%)   
1505                      540.98      1034.13                     0.71%   

     % Price Change [30 Day] % Price Change [YTD] % Price Change [12 Month]  \
1501                  27.82%             (11.70%)                     9.96%   
1502                  29.70%             (22.33%)                   (4.10%)   
1503                  22.04%             (40.49%)                  (36.39%)   
1504                  13.95%              (3.64%)                    25.29%   
1505                  42.98%             (38.36%)                  (16.35%)   

      Price Close   P/E*†  P/BV*†                      Primary Industry  
1501       225.55  23.78x   6.93x  Electronic Equipment and Instruments  
1502       116.25  21.21x   1.93x                 Health Care Equipment  
1503        30.90   9.59x   0.73x                        Regional Banks  
1504       127.53  40.89x  22.39x                       Pharmaceuticals  
1505        21.29   8.00x   1.18x                        Apparel Retail

## 1.2 Clean the Data, Create the DataFrame, and Save CSV File

In [139]:
sp_df.columns

Index(['Company Name', 'Exchange:Ticker', 'Currency',
       'Market Cap (mm) [Latest]*†', 'Revenue (mm)',
       '% Price Change [Last Day]', '% Price Change [30 Day]',
       '% Price Change [YTD]', '% Price Change [12 Month]', 'Price Close',
       'P/E*†', 'P/BV*†', 'Primary Industry'],
      dtype='object')

In [140]:
sp_df.columns = ['company', 'Exchange:Ticker', 'currency',
                 'marketcap_mm', 'revenue_mm', 'pct_price_change_lastday',
                 'pct_pricechange_30day', 'pct_price_change_ytd', 'pct_price_change_12_month', 
                 'price_close', 'price_earnings_ratio', 'price_bookvalue_ratio', 'industry']

In [141]:
sp_df['exchange'] = sp_df['Exchange:Ticker'].apply(lambda i:i.split(":")[0])
sp_df['ticker']  = sp_df['Exchange:Ticker'].apply(lambda i:i.split(":")[1])

In [142]:
sp_df = sp_df.drop(columns=['Exchange:Ticker','currency'])
sp_df.columns

Index(['company', 'marketcap_mm', 'revenue_mm', 'pct_price_change_lastday',
       'pct_pricechange_30day', 'pct_price_change_ytd',
       'pct_price_change_12_month', 'price_close', 'price_earnings_ratio',
       'price_bookvalue_ratio', 'industry', 'exchange', 'ticker'],
      dtype='object')

In [143]:
sp_df = sp_df[['company','ticker', 'price_close', 'pct_price_change_lastday', 
               'pct_pricechange_30day', 'pct_price_change_ytd', 'pct_price_change_12_month',
               'price_earnings_ratio', 'price_bookvalue_ratio', 'marketcap_mm', 
               'revenue_mm','exchange', 'industry']]

In [144]:
pct_price_change_lastday = sp_df['pct_price_change_lastday'].apply(lambda x: x.replace('-', ""))
pct_price_change_lastday = pct_price_change_lastday.apply(lambda x: x.replace('(', '-'))
pct_price_change_lastday = pct_price_change_lastday.apply(lambda x: x.replace(')', ''))
pct_price_change_lastday = pct_price_change_lastday.apply(lambda x: x.replace('%', ''))
pct_price_change_lastday = pd.to_numeric(pct_price_change_lastday)

pct_pricechange_30day = sp_df['pct_pricechange_30day'].apply(lambda x: x.replace('-', ""))
pct_pricechange_30day = pct_pricechange_30day.apply(lambda x: x.replace('(', '-'))
pct_pricechange_30day = pct_pricechange_30day.apply(lambda x: x.replace(')', ''))
pct_pricechange_30day = pct_pricechange_30day.apply(lambda x: x.replace('%', ''))
pct_pricechange_30day = pd.to_numeric(pct_pricechange_30day)

pct_price_change_ytd = sp_df['pct_price_change_ytd'].apply(lambda x: x.replace('-', ""))
pct_price_change_ytd = pct_price_change_ytd.apply(lambda x: x.replace('(', '-'))
pct_price_change_ytd = pct_price_change_ytd.apply(lambda x: x.replace(')', ''))
pct_price_change_ytd = pct_price_change_ytd.apply(lambda x: x.replace('%', ''))
pct_price_change_ytd = pd.to_numeric(pct_price_change_ytd)

pct_price_change_12_month = sp_df['pct_price_change_12_month'].apply(lambda x: x.replace('-', ""))
pct_price_change_12_month = pct_price_change_12_month.apply(lambda x: x.replace('(', '-'))
pct_price_change_12_month = pct_price_change_12_month.apply(lambda x: x.replace(')', ''))
pct_price_change_12_month = pct_price_change_12_month.apply(lambda x: x.replace('%', ''))
pct_price_change_12_month = pd.to_numeric(pct_price_change_12_month)

In [145]:
sp_df['pct_price_change_lastday'] = pct_price_change_lastday
sp_df['pct_pricechange_30day'] = pct_pricechange_30day
sp_df['pct_price_change_ytd'] = pct_price_change_ytd
sp_df['pct_price_change_12_month'] = pct_price_change_12_month

#### Drop the NA rows <br> Notice here: sp_df is the dataframe before dropping NAs, and sp_df_clean is the datafram after dropping NAs.

In [170]:
sp_df_clean = sp_df.dropna()
sp_df_clean.shape

(1498, 13)

**Save final dataframe to csv**

In [182]:
sp_df_clean.to_csv("../../data/SP1500.csv", index = False)

** Have a look at the cleaned dataset **

In [172]:
sp_df_clean.head()

company ticker  price_close  pct_price_change_lastday  \
0   3D Systems Corporation    DDD         7.77                     -8.48   
1               3M Company    MMM       148.60                     -2.19   
2                8x8, Inc.   EGHT        15.89                     -6.31   
3  A. O. Smith Corporation    AOS        41.65                     -1.72   
4               AAON, Inc.   AAON        45.41                     -4.68   

   pct_pricechange_30day  pct_price_change_ytd  pct_price_change_12_month  \
0                  17.37                -11.20                     -26.91   
1                  11.61                -15.77                     -20.14   
2                  23.47                -13.17                     -32.12   
3                  16.96                -12.57                     -20.67   
4                  -3.14                 -8.10                      -8.94   

  price_earnings_ratio price_bookvalue_ratio  marketcap_mm revenue_mm  \
0                   NM                 1.75x        923.60     629.09   
1               17.47x                 8.42x      85474.18      32348   
2                   NM                 7.30x       1628.06     418.53   
3               18.75x                 4.06x       6747.51     2992.7   
4               44.47x                 8.15x       2363.66     469.33   

   exchange                                      industry  
0      NYSE  Technology Hardware, Storage and Peripherals  
1      NYSE                      Industrial Conglomerates  
2      NYSE                          Application Software  
3      NYSE                             Building Products  
4  NasdaqGS                             Building Products

## 1.3 How many constituents and industries are part of the S&P 1500?
### Number of constituents and industries in S&P 1500 - BEFORE dropping NAs

In [173]:
sp_df.shape

(1506, 13)

In [174]:
len(sp_df['industry'].unique())

148

### Number of constituents and industries in S&P 1500 - AFTER dropping NAs

In [175]:
sp_df_clean.shape

(1498, 13)

In [176]:
len(sp_df_clean['industry'].unique())

148

** - Answer : <br> Before dropping NAs, there are 1506 constituents and 148 industries in the S&P 1500. <br> After dropping NAs, there are 1498 constituents and 148 industries in the S&P 1500. **

## 1.4 Determine the top 5 worst & best performing industries

#### 什么是best performance industry的定义？看看后两个能不能有启示 有启示YTD或者CGxx一个指标，或者只看YTD嘛，或者网上查一下标准

In [192]:
sp_industry = sp_df_clean.groupby(['industry'])

<bound method GroupBy.mean of <pandas.core.groupby.groupby.SeriesGroupBy object at 0x7f712ab22780>>

In [198]:
sp_industry['pct_price_change_ytd'].mean().sort_values(ascending=False).head(10)

industry
Gold                                     21.660000
Internet and Direct Marketing Retail     11.883000
Integrated Telecommunication Services    11.860000
Interactive Home Entertainment           11.260000
Internet Services and Infrastructure      8.685000
Food Retail                               4.370000
Health Care Supplies                      4.056667
Hypermarkets and Super Centers            1.935000
Alternative Carriers                      0.167500
Household Products                       -1.132222
Name: pct_price_change_ytd, dtype: float64

In [199]:
sp_industry['pct_pricechange_30day'].mean().sort_values(ascending=False).head(10)

industry
Oil and Gas Exploration and Production    103.465000
Coal and Consumable Fuels                  79.480000
Leisure Facilities                         76.600000
Computer and Electronics Retail            63.102500
Specialized Consumer Services              52.442500
Homefurnishing Retail                      51.273333
Oil and Gas Storage and Transportation     50.851667
Restaurants                                48.979200
Housewares and Specialties                 48.850000
Homebuilding                               45.356250
Name: pct_pricechange_30day, dtype: float64

In [201]:
sp_df_clean.sort_values(by=['pct_price_change_ytd'], ascending=False).head()

company ticker  price_close  \
988   OraSure Technologies, Inc.   OSUR        15.04   
1211             Stamps.com Inc.   STMP       154.49   
458         Enphase Energy, Inc.   ENPH        43.25   
1033        PetMed Express, Inc.   PETS        38.57   
810          Luminex Corporation   LMNX        35.59   

      pct_price_change_lastday  pct_pricechange_30day  pct_price_change_ytd  \
988                      -5.65                  29.88                 87.30   
1211                     -2.39                  24.32                 84.97   
458                      -7.64                  45.28                 65.52   
1033                     -2.53                  36.87                 63.99   
810                      -1.28                  31.57                 53.67   

      pct_price_change_12_month price_earnings_ratio price_bookvalue_ratio  \
988                       56.02               55.48x                 3.02x   
1211                      82.42               45.92x                 3.95x   
458                      233.20               34.89x                19.56x   
1033                      78.98               30.72x                 6.04x   
810                       56.44                   NM                 3.39x   

      marketcap_mm revenue_mm  exchange                              industry  
988         943.73     149.49  NasdaqGS                  Health Care Supplies  
1211       2635.02     571.85  NasdaqGS  Internet and Direct Marketing Retail  
458        5356.55     624.33  NasdaqGM               Semiconductor Equipment  
1033        770.76      274.4  NasdaqGS  Internet and Direct Marketing Retail  
810        1620.45     334.64  NasdaqGS      Life Sciences Tools and Services

## 1.5 Create a new data frame for the worst industry and a new data frame for the best industry both based on % Price Change [YTD]. For each of these new data frames, use the % Price Change [30 Day] to plot the top five ranking for worst and best performance constituents.